contracts/anchor-handler/src/contract.rs

```diff
@@ -0,0 +1,371 @@
+    STATE,
+};
+use protocol_cosmwasm::anchor_handler::{
+    BridgeAddrResponse, ContractAddrResponse, ExecuteMsg, InstantiateMsg, QueryMsg,
+    ResourceIdResponse, UpdateRecordResponse, WhitelistCheckResponse,
+};
+use protocol_cosmwasm::error::ContractError;
+use protocol_cosmwasm::linkable_anchor::ExecuteMsg as LinkableAnchorExecMsg;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cosmwasm-anchor-handler";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+// ChainType info
+pub const COSMOS_CHAIN_TYPE: [u8; 2] = [4, 0]; // 0x0400
+
+pub const MOCK_CHAIN_ID: u64 = 1;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // Validations
+    if !info.funds.is_empty() {
+        return Err(ContractError::UnnecessaryFunds {});
+    }
+
+    if msg.initial_resource_ids.len() != msg.initial_contract_addresses.len() {
+        return Err(ContractError::Std(StdError::GenericErr {
+            msg: "initial_resource_ids and initial_contract_addresses len mismatch".to_string(),
+        }));
+    }
+
+    // Set "state"
+    let bridge_addr = deps.api.addr_validate(&msg.bridge_addr)?;
+    STATE.save(deps.storage, &State { bridge_addr })?;
+
+    // Save the initial mapping of `resource_id => contract_addr`
+    let n = msg.initial_resource_ids.len();
+    for i in 0..n {
+        let resource_id = msg.initial_resource_ids[i];
+        let contract_addr = deps.api.addr_validate(&msg.initial_contract_addresses[i])?;
+        set_resource(deps.storage, resource_id, contract_addr)?;
+    }
+
+    Ok(Response::new().add_attributes(vec![attr("method", "instantiate")]))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        /* ---  Handler common utils --- */
+        ExecuteMsg::SetResource {
+            resource_id,
+            contract_addr,
+        } => exec_set_resource(deps, info, resource_id, contract_addr),
+        ExecuteMsg::MigrateBridge { new_bridge } => migrate_bridge(deps, info, new_bridge),
+        /* ---------------------------- */
+
+        /* ---  Anchor-handler specific execution entries --- */
+        // @notice Proposal execution should be initiated when a proposal is finalized in the Bridge contract.
+        // by a relayer on the deposit's destination chain.
+        //
+        // Params:
+        //   "resource_id": ResourceID corresponding to a particular set of Anchors
+        //   "data": Consists of {sourceChainID}, {leafIndex}, {merkleRoot} all padded to 32 bytes.
+        //
+        // Data passed into the function should be constructed as follows:
```

Comment:
On Substrate we do encode the `call` of the function being executed into the proposal data, so I would be curious if you could provide details on doing that here as well. It might make sense or not, since if we do use the encoded function call directly we'd need to maintain yet another set of proposals for CosmWasm.

Reply:
I think it is okay to follow the `solidity` implementation.
The reason is that 
  - it is good for consistency of system
  - In `substrate`, it also does parsing proposal data. which means no difference in essence.